testsuite/compat/compat.py

```diff
@@ -0,0 +1,301 @@
+test_parser.add_argument(
+    "--tag", "-t", required=True, help="image tag to test, default to HEAD"
+)
+test_parser.add_argument(
+    "--dry", action="store_true", help="dry run without upgrading the cluster"
+)
+test_parser.add_argument(
+    "--workspace", "-w", help="workspace to upgrade, overriding auto-selection",
+)
+test_mode_subparser = test_parser.add_subparsers(
+    required=True, title="test mode commands", dest="test_mode"
+)
+# command: compat test --tag <tag> testnet
+testnet_parser = test_mode_subparser.add_parser(
+    "testnet", help="run compat test against testnet"
+)
+# command: compat test --tag <tag> manual --alt-tag <alt_tag>
+manual_parser = test_mode_subparser.add_parser(
+    "manual", help="run compat test against manually specified revision"
+)
+manual_parser.add_argument(
+    "--alt-tag",
+    required=True,
+    help="alternative image tag to test against specified tag, for debugging",
+)
+args = parser.parse_args()
+
+COMMAND = args.command
+ADMIN_MODE = args.admin_mode if "admin_mode" in args.__dict__ else None
+TEST_MODE = args.test_mode if "test_mode" in args.__dict__ else None
+
+# parse admin
+if COMMAND == "admin":
+    if ADMIN_MODE == "unlock":
+        unlock_cluster(args.cluster_lock)
+        print(f"Successfully unlocked {args.cluster_lock}!")
+    elif ADMIN_MODE == "reset":
+        update_service_task_def(args.cluster, f"{args.cluster}-validator-0", 1)
+        update_service_task_def(args.cluster, f"{args.cluster}-validator-1", 1)
+        update_service_task_def(args.cluster, f"{args.cluster}-validator-2", 1)
+        update_service_task_def(args.cluster, f"{args.cluster}-validator-3", 1)
+        update_service_task_def(args.cluster, f"{args.cluster}-fullnode-0", 1)
+        update_service_task_def(args.cluster, f"{args.cluster}-fullnode-1", 1)
+    sys.exit(0)
+
+
+print("Starting compat test with the following configuration:")
+print(f"\tTESTNET_COMPATIBLE={TESTNET_COMPATIBLE}")
+print(f"\tPREV_COMPATIBLE={PREV_COMPATIBLE}")
+print(f"\tECS_POOL_SIZE={ECS_POOL_SIZE}")
+
+# get the right tags for testing
+TEST_TAG = args.tag
+if TEST_MODE == "testnet":
+    ALT_TEST_TAG = latest_testnet_tag()
+else:
+    ALT_TEST_TAG = args.alt_tag
+
+if not all(
+    [
+        batch_image_exists("libra_validator", [TEST_TAG, ALT_TEST_TAG]),
+        batch_image_exists("libra_safety_rules", [TEST_TAG, ALT_TEST_TAG]),
+    ]
+):
+    print("Missing images, exiting...")
+    sys.exit(1)
+
+if args.workspace:
+    WORKSPACE = args.workspace
+elif args.dry:
+    print("Dry run requires explicit workspace -w")
+    sys.exit(1)
+else:
+    WORKSPACE = select_compat_cluster(ECS_POOL_SIZE, TEST_TAG)
+    # fails if no workspace
+
+def_map = get_task_defs(WORKSPACE)
+
+# get the task definition families as they are the def_map keys
+v0_fam = f"{WORKSPACE}-validator-0"
+v1_fam = f"{WORKSPACE}-validator-1"
+v2_fam = f"{WORKSPACE}-validator-2"
+v3_fam = f"{WORKSPACE}-validator-3"
```

Comment:
Ideally this should support arbitrary sized deployments (even if the size needs to be provided to this script).

Reply:
Agreed @ Python AWS API. I'll hold off on it for now though, until this MVP proves useful enough. Then we can decide if we want to do that or make the better decision of just porting it to Rust :)

Reply:
> Ideally this should support arbitrary sized deployments (even if the size needs to be provided to this script).

So currently the way the upgrades happen for the network are very specific to the network topology of 4 validators and 2 fullnodes. The cluster needs to be first created by Terraform before compat can grab it.

I imagine this tool can also be used to test certain types of upgrades on network with various configurations. They won't be truly "arbitrary" though, as the network would have to be predefined and created via Terraform -- then compat can take care of the rest, applying revision A to N/3 + 1 validators, and B to the rest, etc.

Reply:
I'm fine if the script actually only works with N=4, I think the code is just nicer if it does `for i in range(num_validators)` instead of copy/pasta lines.

Reply:
Gotcha. I originally thought it would be easier to understand if the lines were more explicit, but I've updated it